### ISSUE TITLE
Growing latency on high-traffic blogs

### ISSUE THREAD

**Issue by vadimdemedes** [TryGhost/Ghost#6023]
**TL;DR**

On increased amount of requests (requests per second, next "RPS") Ghost starts to consume a growing amount of memory, latency dramatically increases and requests start to hang.
### Tests

I used [loadtest](https://github.com/alexfernandez/loadtest) tool for load-testing instead of ab or wrk, because it allows to set desired RPS and produces continuous output along with completed requests, errored requests and mean latency.

Before each test run, host server was rebooted.
#### 50 RPS ([view live](https://snapshot.raintank.io/dashboard/snapshot/Le6t6HaVUQfESlk5kWd3JbN5YBR1Bcep)):

<img width="1191" alt="screen shot 2015-10-29 at 1 48 54 pm" src="https://cloud.githubusercontent.com/assets/697676/10818723/1363795c-7e44-11e5-94b6-845a82c31466.png">

```
root@dev:~# loadtest --rps 50
Requests: 0, requests per second: 0, mean latency: 0 ms
Requests: 228, requests per second: 46, mean latency: 140 ms
Requests: 466, requests per second: 48, mean latency: 50 ms
Requests: 728, requests per second: 52, mean latency: 50 ms
Requests: 978, requests per second: 50, mean latency: 40 ms
Requests: 1228, requests per second: 50, mean latency: 30 ms
Requests: 1478, requests per second: 50, mean latency: 30 ms
Requests: 1728, requests per second: 50, mean latency: 60 ms
```

Under 50 RPS everything is fine and works as expected. The latency is low, life is good.
#### 100 RPS ([view live](https://snapshot.raintank.io/dashboard/snapshot/RSgatumXEGQnzOdfedbD3moTw7OJe49O)):

<img width="1186" alt="screen shot 2015-10-29 at 1 53 04 pm" src="https://cloud.githubusercontent.com/assets/697676/10818781/6c080668-7e44-11e5-9bc6-a96dc5a55087.png">

```
root@dev:~# loadtest --rps 100
Requests: 0, requests per second: 0, mean latency: 0 ms
Requests: 235, requests per second: 47, mean latency: 1220 ms
Requests: 526, requests per second: 58, mean latency: 3190 ms
Requests: 857, requests per second: 66, mean latency: 5130 ms
Requests: 1120, requests per second: 53, mean latency: 6920 ms
Requests: 1373, requests per second: 51, mean latency: 9550 ms
Requests: 1618, requests per second: 49, mean latency: 11880 ms
Requests: 1904, requests per second: 57, mean latency: 14310 ms
```

Now things start to look bad. Memory usage is increased, but that's not the issue. The issue is latency, it increases linearly with time and amount of accepted requests.
#### 200 RPS ([view live](https://snapshot.raintank.io/dashboard/snapshot/5V73zXnCle4bkOq5mYdsLd4Yz6FMrpdv))

<img width="1185" alt="screen shot 2015-10-29 at 1 58 41 pm" src="https://cloud.githubusercontent.com/assets/697676/10818889/3095b67e-7e45-11e5-8e4c-9c20e54347c6.png">

```
root@dev:~# loadtest --rps 200
Requests: 0, requests per second: 0, mean latency: 0 ms
Requests: 63, requests per second: 13, mean latency: 2870 ms
Requests: 162, requests per second: 20, mean latency: 6550 ms
Requests: 282, requests per second: 24, mean latency: 11260 ms
Requests: 442, requests per second: 32, mean latency: 15600 ms
Requests: 815, requests per second: 75, mean latency: 9250 ms
Errors: 201, accumulated errors: 201, 24.7% of total requests
Requests: 1815, requests per second: 200, mean latency: 5020 ms
Errors: 785, accumulated errors: 986, 54.3% of total requests
Requests: 2815, requests per second: 200, mean latency: 6720 ms
Errors: 756, accumulated errors: 1742, 61.9% of total requests
```

Requests start to fail and latency is constantly high. 
### Reason for high-latency

High latency yet with average memory consumption could be explained, that Ghost is running some blocking operations during each request. After taking a peek at the heapdump, the most obvious blocking pieces were functions and objects related to Knex and Bookshelf:

<img width="1001" alt="screen shot 2015-10-29 at 2 06 39 pm" src="https://cloud.githubusercontent.com/assets/697676/10819050/565615e2-7e46-11e5-8afa-e1ac589d22c5.png">

MySQL was running at the same server with Ghost (4 GB RAM), so I doubt that MySQL is the bottleneck. 

Initially, I suspected that kernel's socket queue is the reason, but after customizing Ghost code to log new socket connections (and confirming that Ghost did actually receive sockets), this theory was thrown away.

During the initial memory usage testing and the testing you are reading about now, I haven't detected memory leaks. If you take a look of each of the graphs, you see that memory usage drops after requests are closed/aborted. Although there were situations, where memory usage kept growing/dropping even after load-testing has ended. The reason for this is that operations there were initiated during requests (e.g. database requests) were still running.
### What to do next

These are the possible solutions for this issue:
- Cache for MySQL results in Ghost
- Cache for query building in Ghost
- Rate-limiting all requests, so that we can at least display error page, instead of hanging all incoming requests
- Ideas?

We could also repeat the test, but with MySQL cluster configured for performance, but I doubt that it's MySQL's fault.
### Summary

Although this issue relates only to high-traffic blogs (with RPS >= 100), it is a very important one. Once blog hits this wall, there's no easy/quick/cheap fix for that.

In my opinion, one of the 3 proposed solutions should be implemented in the near future. Rate-limiting is the fastest to implement, but not as effective as the first two.

My conclusions may be wrong, so I'd be great if you could point us in the right direction. If anyone wants to take a look, I can upload the heapdumps somewhere. 
##### Environment details

Ghost host server:

RAM: 4 GB
OS: Ubuntu 14.04 x64
ulimit -n: 100000
node: 0.10.40

Client server (executed a load test):

RAM: 1 GB
OS: Ubuntu 14.04 x64
ulimit -n: 100000

**Comment by kevinansfield** [comment]
@vdemedes are you able to also show overall CPU/memory usage and IOWait alongside the memory graphs? This behaviour seems pretty typical to every server app I've worked with before - as the RPS increases the whole system is constrained and once you reach a critical point latency increases on everything.

200 RPS is crazy high to be pushing through to express on a single machine, the proper solution in my mind here is to have a caching mechanism in front of express, be that CloudFlare or apache/nginx/varnish/etc. It seems like we'd be micro-optimising if we were to put limited resources into eeking out marginal performance increases that would only benefit a tiny proportion of our users especially when they are the ones that would more than likely already have a caching solution in place.

**Comment by vdemedes** [comment]
@kevinansfield I could report these values, but this requires sending stats to statsd, so this will need some work. Unfortunately, I won't be able to do this by the end of the week.

200 RPS is indeed high, but same thing happens with 100 RPS. It works perfectly though with 50 RPS. Plus, 200 RPS is not a lot for Node.js apps.

Absolutely, varnish/cloudflare/etc in front of the Ghost app would solve this. But the point of this report is that there is a need of a caching mechanism inside Ghost. Or at least an error page, explaining that there's a high amount of traffic and this page could not be served. Something friendlier, than infinite load and blank page.

Caching is an essential part of production apps, especially when we're dealing with SQL.

I agree with you on the stuff you mentioned, but the question I had in mind is how to protect from rather than avoid this situation ;)

**Comment by ErisDS** [comment]
I agree that this is a case where, putting a cache in front of Ghost is far more expected than throwing >50rps at it.

However, it looks to me like this is hinting at some interesting structural concerns with the way we depend on bookshelf and knex that are likely to become more apparent as we open up the API. 

We have talked before about having a caching layer inside the API, and improving the caching headers on responses. That need is going to get greater. 

Perhaps we could also look into memoizing query builder objects, so we use the same object to run the same query? Not sure how or where this would work, whether it would need to be in Ghost, or in bookshelf or even in knex, but particularly with the advanced 'browse' queries, I could see us having a query cache, rather than a response cache.

**Comment by ErisDS** [comment]
Closing this for now. The underlying issue here is 1) more of a problem with knex/bookshelf and 2) an edgecase. There are improvements we could make but nothing urgent.
